lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,91 @@
+class LRUHash
+  # This LRU (Least Recently Used) hash will allow
+  # the cleaning of resources as time goes.
+  # The  goal is to remove the least recently used resources
+  # everytime we set a new resource. A default window of
+  # 5 minutes will allow empty item to stay in the hash
+  # for a maximum of 5 minutes
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?
+  attr_reader :table, :minimum_time_in_lru
+  MINIMUM_TIME_IN_LRU = 300
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+
+    def locked?
+      false
+    end
+  end
+
+  [:keys, :clear].each do |attribute|
+    define_method :"#{attribute}" do
+      @lock.synchronize { @table.public_send(attribute) }
+    end
+  end
+
+  def initialize
+    @minimum_time_in_lru = MINIMUM_TIME_IN_LRU
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
+    delete(key)
+    @lock.synchronize do
+      @table[key] = resource
+    end
+    clear_unused_resources
+    resource.updated_at = Time.now
+  end
+
+  def get(key)
+    found = delete(key)
```

Comment:
Some as above. Can you explain your reasoning behind this lock placement being correct?

Reply:
I've added a lock on the `delete` method since it's a public method and I want to avoid race condition there, so I did not want to lock inside the lock but now that I think about it. I can fix that by simply calling `@table.delete(key)` here, I'll get that fixed!